File: 286-walls-and-gates/walls-and-gates.java
class Solution {

    private static final List<int[]> DIRECTIONS = Arrays.asList(
        new int[] {-1, 0},
        new int[] {0, 1},
        new int[] {1, 0},
        new int[] {0, -1}
    );

    public void wallsAndGates(int[][] rooms) {
        // multi source bfs
        int rows = rooms.length;
        int cols = rooms[0].length;
        int EMTPY = Integer.MAX_VALUE;

        Queue<int[]> queue = new LinkedList<>();
        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) {
                if (rooms[r][c] == 0) {
                    queue.add(new int[]{r, c});
                }
            }
        }
        
        while (!queue.isEmpty()) {
            int[] point = queue.poll();
            int r = point[0];
            int c = point[1];
            for (int[] direction : DIRECTIONS) {
                int nr = r + direction[0];
                int nc = c + direction[1];

                if (
                    nr < 0 || nr >= rows || nc < 0 || nc >= cols ||
                    rooms[nr][nc] != EMTPY
                ) {
                    continue;
                }
                rooms[nr][nc] = rooms[r][c] + 1;
                queue.add(new int[] {nr, nc});
            }
        }
    }
}

File: 286-walls-and-gates/walls-and-gates.py
class Solution:
    def wallsAndGates(self, rooms: List[List[int]]) -> None:
        """
        Do not return anything, modify rooms in-place instead.
        """
        rows, cols = len(rooms), len(rooms[0])
        
        q = deque()
        visited = set()
        for r in range(rows):
            for c in range(cols):
                if rooms[r][c] == 0:
                    q.append((r, c))
                    visited.add((r, c))

        dist = 0
        while q:
            dist += 1
            level = len(q)
            for _ in range(level):
                r, c = q.popleft()
                for dr, dc in [(0, 1), (1, 0), (0, -1), (-1, 0)]:
                    nr, nc = r + dr, c + dc
                    if (
                        nr < 0 or nr >= rows or
                        nc < 0 or nc >= cols or
                        (nr, nc) in visited or
                        rooms[nr][nc] == -1
                    ):
                        continue
                    visited.add((nr, nc))
                    q.append((nr, nc))
                    rooms[nr][nc] = dist